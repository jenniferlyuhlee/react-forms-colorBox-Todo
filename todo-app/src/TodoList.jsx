import React, { useState } from "react";
import { v4 as uuid } from "uuid";
import NewTodoForm from "./NewTodoForm"; 
import Todo from "./Todo";

function TodoList() {
    const INITIAL_STATE = []

    const [todoList, setTodoList] = useState(INITIAL_STATE);

    const addTodo = (newTodo) => {
        setTodoList(todoList => [...todoList, {...newTodo, id:uuid()}])
    }

    const removeTodo = (id) => {
        setTodoList(todoList.filter(t => t.id != id));
    }

    return (
        <div>
            <h1>Todo List</h1>
            <NewTodoForm addTodo={addTodo} />
            <div className="Todos">
                {todoList.map(({task, id}) => <Todo
                                            task = {task}
                                            key = {id}
                                            removeTodo={() => removeTodo(id)}/>)}
            </div>
        </div>
    )
}

export default TodoList;