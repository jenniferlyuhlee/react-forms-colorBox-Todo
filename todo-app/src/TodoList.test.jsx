import React from "react";
import { fireEvent, render } from "@testing-library/react"
import TodoList from "./TodoList";
import { expect } from "vitest";

// smoke test
test("renders without crashing", () => {
    render(<TodoList />)
});

// snapshot test
test("matches snapshot", () => {
    const {asFragment} = render(<TodoList />)
    expect(asFragment()).toMatchSnapshot()
})

// specialized tests - add/remove Todo functionality
// mock addTodo
function addTodo(todoList) {
    const taskInput = todoList.getByLabelText("I need to:");
    // fill out form
    fireEvent.change(taskInput, {target: {value: "Test task"}});
    // submit form
    const btn = todoList.getByText("Add!")
    fireEvent.click(btn)
}

test("should add new todo", () => {
    const todoList = render(<TodoList />)
    addTodo(todoList)

    expect(todoList.container.querySelector('.Todo-group')).toBeInTheDocument();

    const deleteBtn = todoList.getByText("Delete");
    expect(deleteBtn).toBeInTheDocument();
    const task = deleteBtn.previousSibling;
    expect(task).toHaveClass("Todo-task");
    expect(task).toHaveTextContent("Test task");
    
})

test("should delete a todo", () => {
    const todoList = render(<TodoList />)
    addTodo(todoList)

    const deleteBtn = todoList.getByText("Delete")
    const task = deleteBtn.previousSibling;
    expect(task).toBeInTheDocument();

    //click to remove Todo
    fireEvent.click(deleteBtn);
    expect(deleteBtn).not.toBeInTheDocument()
    expect(task).not.toBeInTheDocument()
})