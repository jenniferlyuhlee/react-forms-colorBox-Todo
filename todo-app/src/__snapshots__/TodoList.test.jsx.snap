// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Todo List
    </h1>
    <form
      class="NewTodoForm"
    >
      <label
        class="todo-label"
        for="task"
      >
        I need to: 
      </label>
      <input
        class="todo-input"
        id="task"
        name="task"
        placeholder="Ex: Wash dishes"
        type="text"
        value=""
      />
      <button
        class="NewTodoForm-btn"
      >
        Add!
      </button>
    </form>
    <div
      class="Todos"
    />
  </div>
</DocumentFragment>
`;
